File: ios/Sources/ForgeIOSRuntime/MarkdownFenceParser.swift
import Foundation

public enum MarkdownFencePart: Equatable, Sendable {
    case text(String)
    case fence(raw: String, language: String, header: String, body: String, closed: Bool)
}

/// Framework-neutral markdown fence segmentation for Forge hosts.
public enum MarkdownFenceParser {
    private static let marker = "```"

    public static func parse(_ content: String) -> [MarkdownFencePart] {
        guard !content.isEmpty else { return [] }
        var parts: [MarkdownFencePart] = []
        var cursor = content.startIndex
        while cursor < content.endIndex {
            guard let opening = content.range(of: marker, range: cursor..<content.endIndex) else {
                appendText(&parts, String(content[cursor...]))
                break
            }
            if opening.lowerBound > cursor { appendText(&parts, String(content[cursor..<opening.lowerBound])) }
            let languageStart = opening.upperBound
            var languageEnd = languageStart
            while languageEnd < content.endIndex, isLanguageCharacter(content[languageEnd]) {
                languageEnd = content.index(after: languageEnd)
            }
            guard let envelope = headerAndBodyStart(in: content, after: languageEnd) else {
                appendText(&parts, marker)
                cursor = languageStart
                continue
            }
            let language = String(content[languageStart..<languageEnd]).lowercased()
            guard let closing = closingMarker(in: content, from: envelope.bodyStart) else {
                parts.append(.fence(raw: String(content[opening.lowerBound...]), language: language, header: envelope.header, body: String(content[envelope.bodyStart...]), closed: false))
                break
            }
            parts.append(.fence(raw: String(content[opening.lowerBound..<closing.upperBound]), language: language, header: envelope.header, body: String(content[envelope.bodyStart..<closing.lowerBound]), closed: true))
            cursor = closing.upperBound
        }
        return parts
    }

    private struct HeaderAndBodyStart { let header: String; let bodyStart: String.Index }
    private static func headerAndBodyStart(in content: String, after index: String.Index) -> HeaderAndBodyStart? {
        guard index < content.endIndex else { return nil }
        switch content[index] {
        case "\n": return HeaderAndBodyStart(header: "", bodyStart: content.index(after: index))
        case "\r":
            let next = content.index(after: index)
            return next < content.endIndex && content[next] == "\n" ? HeaderAndBodyStart(header: "", bodyStart: content.index(after: next)) : nil
        case "{", "[": return HeaderAndBodyStart(header: "", bodyStart: index)
        case " ", "\t":
            var newline = index
            while newline < content.endIndex, content[newline] != "\n", content[newline] != "\r" { newline = content.index(after: newline) }
            guard newline < content.endIndex else { return nil }
            if content[newline] == "\n" { return HeaderAndBodyStart(header: String(content[index..<newline]).trimmingCharacters(in: .whitespaces), bodyStart: content.index(after: newline)) }
            let next = content.index(after: newline)
            guard next < content.endIndex, content[next] == "\n" else { return nil }
            return HeaderAndBodyStart(header: String(content[index..<newline]).trimmingCharacters(in: .whitespaces), bodyStart: content.index(after: next))
        default: return nil
        }
    }
    private static func isLanguageCharacter(_ character: Character) -> Bool { character.isLetter || character.isNumber || character == "_" || character == "+" || character == "-" }

    private static func closingMarker(in content: String, from bodyStart: String.Index) -> Range<String.Index>? {
        var searchStart = bodyStart
        while let candidate = content.range(of: marker, range: searchStart..<content.endIndex) {
            let atLineStart = candidate.lowerBound == content.startIndex || content[content.index(before: candidate.lowerBound)] == "\n"
            let body = String(content[bodyStart..<candidate.lowerBound])
            // Compact JSON fences are emitted without line breaks. Only accept
            // their closing marker once the full body is valid JSON.
            if atLineStart || isValidJSON(body) {
                return candidate
            }
            searchStart = candidate.upperBound
        }
        return nil
    }

    private static func isValidJSON(_ value: String) -> Bool {
        let data = Data(value.trimmingCharacters(in: .whitespacesAndNewlines).utf8)
        return !data.isEmpty && (try? JSONSerialization.jsonObject(with: data)) != nil
    }

    private static func appendText(_ parts: inout [MarkdownFencePart], _ value: String) {
        guard !value.isEmpty else { return }
        if case .text(let existing)? = parts.last { parts[parts.count - 1] = .text(existing + value) } else { parts.append(.text(value)) }
    }
}

/// Parses standard key=value attributes from a fence header without regexes.
public enum MarkdownFenceHeader {
    public static func attributes(_ header: String) -> [String: String] {
        var result: [String: String] = [:]
        var cursor = header.startIndex
        while cursor < header.endIndex {
            while cursor < header.endIndex, header[cursor].isWhitespace { cursor = header.index(after: cursor) }
            let keyStart = cursor
            while cursor < header.endIndex, header[cursor].isLetter || header[cursor].isNumber || header[cursor] == "_" || header[cursor] == "-" { cursor = header.index(after: cursor) }
            guard keyStart < cursor, cursor < header.endIndex, header[cursor] == "=" else { return result }
            let key = String(header[keyStart..<cursor]).lowercased()
            cursor = header.index(after: cursor)
            let value: String
            if cursor < header.endIndex, (header[cursor] == "\"" || header[cursor] == "'") {
                let quote = header[cursor]; cursor = header.index(after: cursor); let start = cursor
                while cursor < header.endIndex, header[cursor] != quote { cursor = header.index(after: cursor) }
                guard cursor < header.endIndex else { return result }
                value = String(header[start..<cursor]); cursor = header.index(after: cursor)
            } else {
                let start = cursor
                while cursor < header.endIndex, !header[cursor].isWhitespace { cursor = header.index(after: cursor) }
                value = String(header[start..<cursor])
            }
            result[key] = value
        }
        return result
    }
}

File: android/sdk/src/main/java/com/viant/forgeandroid/ui/MarkdownFenceParser.kt
package com.viant.forgeandroid.ui

import kotlinx.serialization.json.Json

/**
 * Framework-neutral markdown fence segmentation used by hosts before they
 * decide which Forge surface, if any, owns a fenced payload.
 */
sealed interface MarkdownFencePart {
    data class Text(val value: String) : MarkdownFencePart

    data class Fence(
        val raw: String,
        val language: String,
        val header: String,
        val body: String,
        val closed: Boolean
    ) : MarkdownFencePart
}

/**
 * A small scanner instead of a regular expression so streamed, malformed, and
 * adjacent fences retain their original text without backtracking ambiguity.
 */
object MarkdownFenceParser {
    private const val marker = "```"
    private val json = Json { ignoreUnknownKeys = true }

    fun parse(content: String): List<MarkdownFencePart> {
        if (content.isEmpty()) {
            return emptyList()
        }
        val parts = mutableListOf<MarkdownFencePart>()
        var cursor = 0
        while (cursor < content.length) {
            val opening = content.indexOf(marker, cursor)
            if (opening < 0) {
                appendText(parts, content.substring(cursor))
                break
            }
            if (opening > cursor) {
                appendText(parts, content.substring(cursor, opening))
            }
            val languageStart = opening + marker.length
            var languageEnd = languageStart
            while (languageEnd < content.length && isLanguageCharacter(content[languageEnd])) {
                languageEnd++
            }
            val headerAndBody = headerAndBodyStart(content, languageEnd)
            if (headerAndBody == null) {
                // This marker is not a valid fence opener. Keep it verbatim.
                appendText(parts, marker)
                cursor = languageStart
                continue
            }
            val closing = closingMarker(content, headerAndBody.bodyStart)
            if (closing < 0) {
                parts += MarkdownFencePart.Fence(
                    raw = content.substring(opening),
                    language = content.substring(languageStart, languageEnd).lowercase(),
                    header = headerAndBody.header,
                    body = content.substring(headerAndBody.bodyStart),
                    closed = false
                )
                break
            }
            parts += MarkdownFencePart.Fence(
                raw = content.substring(opening, closing + marker.length),
                language = content.substring(languageStart, languageEnd).lowercase(),
                header = headerAndBody.header,
                body = content.substring(headerAndBody.bodyStart, closing),
                closed = true
            )
            cursor = closing + marker.length
        }
        return parts
    }

    private data class HeaderAndBodyStart(val header: String, val bodyStart: Int)

    private fun closingMarker(content: String, bodyStart: Int): Int {
        var searchStart = bodyStart
        while (searchStart < content.length) {
            val candidate = content.indexOf(marker, searchStart)
            if (candidate < 0) {
                return -1
            }
            val atLineStart = candidate == 0 || content[candidate - 1] == '\n' || content[candidate - 1] == '\r'
            val body = content.substring(bodyStart, candidate).trim()
            if (atLineStart || isValidJson(body)) {
                return candidate
            }
            searchStart = candidate + marker.length
        }
        return -1
    }

    private fun isValidJson(value: String): Boolean {
        if (value.isEmpty()) {
            return false
        }
        return runCatching { json.parseToJsonElement(value) }.isSuccess
    }

    private fun headerAndBodyStart(content: String, languageEnd: Int): HeaderAndBodyStart? {
        if (languageEnd >= content.length) {
            return null
        }
        return when (content[languageEnd]) {
            '\n' -> HeaderAndBodyStart("", languageEnd + 1)
            '\r' -> if (languageEnd + 1 < content.length && content[languageEnd + 1] == '\n') HeaderAndBodyStart("", languageEnd + 2) else null
            '{', '[' -> HeaderAndBodyStart("", languageEnd)
            ' ', '\t' -> headerLine(content, languageEnd)
            else -> null
        }
    }

    private fun headerLine(content: String, start: Int): HeaderAndBodyStart? {
        var newline = start
        while (newline < content.length && content[newline] != '\n' && content[newline] != '\r') {
            newline++
        }
        if (newline == content.length) {
            return null
        }
        val bodyStart = when (content[newline]) {
            '\n' -> newline + 1
            '\r' -> if (newline + 1 < content.length && content[newline + 1] == '\n') newline + 2 else return null
            else -> return null
        }
        return HeaderAndBodyStart(content.substring(start, newline).trim(), bodyStart)
    }

    private fun isLanguageCharacter(value: Char): Boolean {
        return value.isLetterOrDigit() || value == '_' || value == '+' || value == '-'
    }

    private fun appendText(parts: MutableList<MarkdownFencePart>, value: String) {
        if (value.isEmpty()) {
            return
        }
        val previous = parts.lastOrNull()
        if (previous is MarkdownFencePart.Text) {
            parts[parts.lastIndex] = previous.copy(value = previous.value + value)
        } else {
            parts += MarkdownFencePart.Text(value)
        }
    }
}

/** Parses standard key=value attributes from a fence header without regexes. */
object MarkdownFenceHeader {
    fun attributes(header: String): Map<String, String> {
        val attributes = linkedMapOf<String, String>()
        var cursor = 0
        while (cursor < header.length) {
            while (cursor < header.length && header[cursor].isWhitespace()) cursor++
            val keyStart = cursor
            while (cursor < header.length && (header[cursor].isLetterOrDigit() || header[cursor] == '_' || header[cursor] == '-')) cursor++
            if (keyStart == cursor || cursor >= header.length || header[cursor] != '=') {
                return attributes
            }
            val key = header.substring(keyStart, cursor).lowercase()
            cursor++
            val valueStart = cursor
            val value = if (cursor < header.length && (header[cursor] == '\"' || header[cursor] == '\'')) {
                val quote = header[cursor++]
                val quotedStart = cursor
                while (cursor < header.length && header[cursor] != quote) cursor++
                if (cursor >= header.length) return attributes
                header.substring(quotedStart, cursor++)
            } else {
                while (cursor < header.length && !header[cursor].isWhitespace()) cursor++
                header.substring(valueStart, cursor)
            }
            attributes[key] = value
        }
        return attributes
    }
}
